File: wagerproof-ios-native/Wagerproof/Features/Agents/Creation/Step5CustomInsightsView.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels
import WagerproofStores

/// Step 5 of the wizard — free-form custom insight fields. Ports
/// `components/agents/creation/Screen5_CustomInsights.tsx`.
///
/// Converted to native Form/Section — each insight field is its own Section
/// containing a multi-line TextField (axis: .vertical). The collapsible
/// InsightCard behaviour is replaced by always-visible sections (Form itself
/// groups them clearly). Char limits (500/500/300/300), filled-badge logic,
/// and bindings to `store.draft.customInsights.*` are all preserved verbatim.
struct Step5CustomInsightsView: View {
    @Bindable var store: AgentCreationStore

    private struct FieldConfig {
        let title: String
        let icon: String
        let description: String
        let placeholder: String
        let maxLength: Int
        let keyPath: WritableKeyPath<AgentCustomInsights, String?>
    }

    private let fields: [FieldConfig] = [
        .init(
            title: "Betting Philosophy",
            icon: "book.fill",
            description: "Describe your overall approach to betting. What principles guide your decisions?",
            placeholder: "e.g., I believe in value betting and only taking plays where I have a significant edge over the market...",
            maxLength: 500,
            keyPath: \.bettingPhilosophy
        ),
        .init(
            title: "Perceived Edges",
            icon: "chart.line.uptrend.xyaxis",
            description: "What unique insights or edges do you think you have?",
            placeholder: "e.g., I'm particularly good at spotting mispriced totals in divisional games, especially when weather is a factor...",
            maxLength: 500,
            keyPath: \.perceivedEdges
        ),
        .init(
            title: "Situations to Avoid",
            icon: "xmark.octagon",
            description: "What types of games or situations should your agent avoid?",
            placeholder: "e.g., Never bet on primetime games, avoid teams coming off emotional wins, skip games with uncertain QB situations...",
            maxLength: 300,
            keyPath: \.avoidSituations
        ),
        .init(
            title: "Target Situations",
            icon: "target",
            description: "What types of games or situations should your agent prioritize?",
            placeholder: "e.g., Look for home underdogs off a bye week, target early season totals before lines adjust, focus on late-season division games...",
            maxLength: 300,
            keyPath: \.targetSituations
        ),
    ]

    private var filledCount: Int {
        var n = 0
        if (store.draft.customInsights.bettingPhilosophy ?? "").isEmpty == false { n += 1 }
        if (store.draft.customInsights.perceivedEdges ?? "").isEmpty == false { n += 1 }
        if (store.draft.customInsights.avoidSituations ?? "").isEmpty == false { n += 1 }
        if (store.draft.customInsights.targetSituations ?? "").isEmpty == false { n += 1 }
        return n
    }

    var body: some View {
        Form {
            // Intro Section — description + filled count.
            Section {
                EmptyView()
            } header: {
                VStack(alignment: .leading, spacing: 6) {
                    Text("Custom Insights")
                        .font(.system(size: 20, weight: .bold))
                        .foregroundStyle(Color.appTextPrimary)
                        .textCase(nil)
                    Text("Help your agent understand your unique betting perspective. These fields are optional but can improve how well your agent matches your style.")
                        .font(.system(size: 13))
                        .foregroundStyle(Color.appTextSecondary)
                        .textCase(nil)
                    Text("\(filledCount) of \(fields.count) completed (optional)")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(Color.appTextSecondary)
                        .textCase(nil)
                        .padding(.bottom, 4)
                }
                .padding(.bottom, 6)
            }

            // One Section per insight field.
            ForEach(Array(fields.enumerated()), id: \.offset) { _, config in
                insightSection(config: config, binding: optionalStringBinding(for: config.keyPath))
            }

            // Info footer row.
            Section {
                HStack(alignment: .top, spacing: 10) {
                    Image(systemName: "info.circle")
                        .font(.system(size: 16))
                        .foregroundStyle(Color.appTextSecondary)
                    Text("These insights help the AI understand your betting philosophy. You can always edit them later from the agent settings.")
                        .font(.system(size: 13))
                        .foregroundStyle(Color.appTextSecondary)
                        .fixedSize(horizontal: false, vertical: true)
                }
                .listRowBackground(Color.clear)
            }
        }
    }

    @ViewBuilder
    private func insightSection(config: FieldConfig, binding: Binding<String>) -> some View {
        let charCount = binding.wrappedValue.count
        let isOverLimit = charCount > config.maxLength
        let isFilled = !binding.wrappedValue.isEmpty

        Section {
            // Multi-line TextField — axis:.vertical grows with content.
            TextField(config.placeholder, text: binding, axis: .vertical)
                .font(.system(size: 15))
                .lineLimit(4...10)
        } header: {
            HStack(spacing: 6) {
                Image(systemName: config.icon)
                    .font(.system(size: 13))
                    .foregroundStyle(isFilled ? Color(hex: 0x00E676) : Color.secondary)
                Text(config.title)
                    .textCase(nil)
                if isFilled {
                    // "Filled" pill moved to header to match the original badge.
                    Text("Filled")
                        .font(.system(size: 10, weight: .semibold))
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .foregroundStyle(Color(hex: 0x00E676))
                        .background(Capsule().fill(Color(hex: 0x00E676).opacity(0.18)))
                }
            }
        } footer: {
            VStack(alignment: .leading, spacing: 2) {
                Text(config.description)
                    .font(.system(size: 12))
                HStack {
                    Spacer()
                    Text("\(charCount)/\(config.maxLength)")
                        .font(.system(size: 11))
                        .foregroundStyle(isOverLimit ? Color(hex: 0xEF4444) : Color.secondary)
                }
            }
        }
    }

    private func optionalStringBinding(for keyPath: WritableKeyPath<AgentCustomInsights, String?>) -> Binding<String> {
        Binding(
            get: { store.draft.customInsights[keyPath: keyPath] ?? "" },
            set: { newValue in
                store.draft.customInsights[keyPath: keyPath] = newValue.isEmpty ? nil : newValue
            }
        )
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/creation/Step5CustomInsightsView.kt
package com.wagerproof.app.features.agents.creation

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.app.features.agents.creation.inputs.clickableNoRipple
import com.wagerproof.core.design.icons.AppIcon
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.AgentCustomInsights
import com.wagerproof.core.stores.AgentCreationStore

private enum class InsightField { PHILOSOPHY, EDGES, AVOID, TARGET }
private data class InsightConfig(
    val field: InsightField,
    val title: String,
    val icon: String,
    val description: String,
    val placeholder: String,
    val maxLength: Int,
)

private val InsightFields = listOf(
    InsightConfig(InsightField.PHILOSOPHY, "Betting Philosophy", "book.fill", "Describe the principles that guide your decisions.", "e.g., I believe in value betting and only take plays with a significant market edge...", 500),
    InsightConfig(InsightField.EDGES, "Perceived Edges", "chart.line.uptrend.xyaxis", "What unique insights or edges do you think you have?", "e.g., I'm particularly good at spotting mispriced totals in divisional games...", 500),
    InsightConfig(InsightField.AVOID, "Situations to Avoid", "xmark.octagon", "What games or situations should your agent avoid?", "e.g., Avoid teams coming off emotional wins and games with uncertain QB situations...", 300),
    InsightConfig(InsightField.TARGET, "Target Situations", "target", "What games or situations should your agent prioritize?", "e.g., Look for home underdogs off a bye and late-season division games...", 300),
)

/** Step 5: optional long-form steering instructions. */
@Composable
fun Step5CustomInsightsView(store: AgentCreationStore, modifier: Modifier = Modifier) {
    val insights = store.draft.customInsights
    fun value(field: InsightField): String = when (field) {
        InsightField.PHILOSOPHY -> insights.bettingPhilosophy.orEmpty()
        InsightField.EDGES -> insights.perceivedEdges.orEmpty()
        InsightField.AVOID -> insights.avoidSituations.orEmpty()
        InsightField.TARGET -> insights.targetSituations.orEmpty()
    }
    fun update(field: InsightField, raw: String) {
        val text = raw.ifEmpty { null }
        val next = when (field) {
            InsightField.PHILOSOPHY -> insights.copy(bettingPhilosophy = text)
            InsightField.EDGES -> insights.copy(perceivedEdges = text)
            InsightField.AVOID -> insights.copy(avoidSituations = text)
            InsightField.TARGET -> insights.copy(targetSituations = text)
        }
        store.draft = store.draft.copy(customInsights = next)
    }
    val filled = InsightFields.count { value(it.field).isNotEmpty() }

    Column(
        modifier.fillMaxWidth().background(AppColors.appSurface).verticalScroll(rememberScrollState()).padding(horizontal = 16.dp, vertical = 16.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Text("Custom Insights", color = AppColors.appTextPrimary, fontSize = 20.sp, fontWeight = FontWeight.Bold, modifier = Modifier.padding(start = 4.dp, top = 8.dp))
        Text("Help your agent understand your betting perspective. These are optional, but improve how closely it matches your style.", color = AppColors.appTextSecondary, fontSize = 13.sp, modifier = Modifier.padding(horizontal = 4.dp))
        Text("$filled of ${InsightFields.size} completed (optional)", color = AppColors.appTextSecondary, fontSize = 12.sp, fontWeight = FontWeight.Medium, modifier = Modifier.padding(horizontal = 4.dp, vertical = 4.dp))

        InsightFields.forEach { config ->
            val text = value(config.field)
            val over = text.length > config.maxLength
            Row(Modifier.fillMaxWidth().padding(start = 4.dp, top = 18.dp, bottom = 8.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(7.dp)) {
                Icon(AppIcon.fromSystemName(config.icon)?.imageVector ?: AppIcon.SPARKLES.imageVector, null, tint = if (text.isNotEmpty()) AppColors.brandGreenBright else AppColors.appTextSecondary)
                Text(config.title, color = AppColors.appTextSecondary, fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
                if (text.isNotEmpty()) Text("Filled", color = AppColors.brandGreenBright, fontSize = 10.sp, fontWeight = FontWeight.SemiBold, modifier = Modifier.background(AppColors.brandGreenBright.copy(alpha = 0.18f), RoundedCornerShape(50)).padding(horizontal = 7.dp, vertical = 3.dp))
            }
            WizardSectionCard {
                TextField(
                    value = text,
                    onValueChange = { update(config.field, it) },
                    modifier = Modifier.fillMaxWidth().semantics { contentDescription = config.title },
                    placeholder = { Text(config.placeholder, color = AppColors.appTextSecondary.copy(alpha = 0.65f)) },
                    minLines = 4,
                    maxLines = 10,
                    colors = TextFieldDefaults.colors(
                        focusedContainerColor = Color.Transparent,
                        unfocusedContainerColor = Color.Transparent,
                        focusedIndicatorColor = AppColors.brandGreenBright,
                        unfocusedIndicatorColor = Color.Transparent,
                        focusedTextColor = AppColors.appTextPrimary,
                        unfocusedTextColor = AppColors.appTextPrimary,
                    ),
                )
            }
            Row(Modifier.fillMaxWidth().padding(horizontal = 4.dp, vertical = 5.dp)) {
                Text(config.description, color = AppColors.appTextSecondary, fontSize = 12.sp, modifier = Modifier.weight(1f))
                Text("${text.length}/${config.maxLength}", color = if (over) AppColors.appLoss else AppColors.appTextSecondary, fontSize = 11.sp)
            }
        }

        Row(Modifier.fillMaxWidth().padding(vertical = 18.dp), verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
            Icon(AppIcon.INFO_CIRCLE.imageVector, null, tint = AppColors.appTextSecondary)
            Text("These insights help the AI understand your philosophy. You can edit them later in agent settings.", color = AppColors.appTextSecondary, fontSize = 13.sp)
        }
        Spacer(Modifier.height(18.dp))
    }
}
